flowinfra: create flow runner waiter goroutine as a task

yuzefovich opened the pull request:
**distsql: always cancel running queries after flow drain wait time**

This commit removes a cluster setting that determines whether the flows
still running when `server.shutdown.query_wait` elapses are canceled. In
22.2 we introduced this setting with a default value of `true` (meaning
to cancel long-running queries after the wait period) out of caution,
but now it should be safe to remove this escape hatch and always cancel
those stragglers - this was always the intention after all.

Release note: None

**flowinfra: create flow runner waiter goroutine as a task**

This commit makes the change to create goroutine that is spun up to wait
for the remote flow to finish and to clean up after it as a task on the
stopper. This change allows us to fix the possible race between the server
quiescing (which closes the temp engine) and the vectorized flow cleanup
(which might delete its temp storage directory).

Some care had to be taken to prevent some types of races: on one hand, we
cannot start the flow until the new async waiter task is spun up because
in that case the bug this commit is fixing is still present; on the
other hand, the async task cannot block on `Flow.Wait` method until the
flow is started successfully. Such a setup requires coordination between
the new async task and the flow being started in the current goroutine.

Fixes: #92504.
Fixes: #92903.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/93214)
<!-- Reviewable:end -->

nicktrav commented:
Realize this is still WIP, but just wanted to point this out:

```
$ ./dev test ./pkg/sql/execstats --filter 'TestTraceAnalyzer$' --stress
...
390 runs so far, 0 failures, over 45s
...
goroutine 3147 [running]:
github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).GetID(0xc00353f6d0?)
        github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:326 +0x6
github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*RemoteFlowRunner).RunFlow.func1.2()
        github.com/cockroachdb/cockroach/pkg/sql/flowinfra/remote_flow_runner.go:115 +0xae
github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*RemoteFlowRunner).RunFlow.func1.3()
        github.com/cockroachdb/cockroach/pkg/sql/flowinfra/remote_flow_runner.go:127 +0x38
created by github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*RemoteFlowRunner).RunFlow.func1
        github.com/cockroachdb/cockroach/pkg/sql/flowinfra/remote_flow_runner.go:125 +0x2b2
```

yuzefovich commented:
I think the failure that you saw (and what I reproduced locally) was a nil pointer, now fixed. Stressing it myself.

yuzefovich committed:
distsql: always cancel running queries after flow drain wait time

This commit removes a cluster setting that determines whether the flows
still running when `server.shutdown.query_wait` elapses are canceled. In
22.2 we introduced this setting with a default value of `true` (meaning
to cancel long-running queries after the wait period) out of caution,
but now it should be safe to remove this escape hatch and always cancel
those stragglers - this was always the intention after all.

Epic: None

Release note: None

yuzefovich commented:
@knz could you please take a look at the first commit?

@nicktrav this seems to fix the issue:
```
17474 runs so far, 0 failures, over 30m0s
```

knz reviewed (commented):
@andreimatei could you help have a look at this? How do you feel about relying on the order of closers? AFAIK we've never done that before.

> Reviewed 2 of 3 files at r2, 3 of 3 files at r3, 5 of 5 files at r5, 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93214#-:-NIn1mp9Ezc5vRGmRAd2:bvhf175)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach and @yuzefovich)

___
*[`-- commits` line 15 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/93214#-NIn1V29E6qH06N8PmrO):*
"canceled to be always canceled" - what does this mean?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93214)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach and @knz)

___
*[`-- commits` line 15 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/93214#-NIn1V29E6qH06N8PmrO):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

"canceled to be always canceled" - what does this mean?
</blockquote></details>

Indeed, poor phrasing, changed.


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93214)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach, @knz, and @yuzefovich)

___
*[`-- commits` line 39 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/93214#-NIniO4y6r7JHrdhTWFV):*
How were the deletions done before this patch? They somehow survived stopper quiescing?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93214)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @cucaroach, and @knz)

___
*[`-- commits` line 39 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/93214#-NIniO4y6r7JHrdhTWFV):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

How were the deletions done before this patch? They somehow survived stopper quiescing?
</blockquote></details>

AFAIK the flow infrastructure is completely unaware of the concept of quiescing, so we - generally speaking - get lucky that the quiescing process takes enough time for all remote flows to perform the cleanup, including performing the deletion of the temp storage directories. Also we probably don't have enough tests that force the spilling to disk (if we don't spill, then there is nothing to delete) so that the cases when we don't get lucky are quite rare. Yet, under stress they can be reproduced pretty easily. It also seems possible that the likelihood of the flakes has recently increased with us fully parallelizing the setup of all flows (until #89649, remote flows had a head start).


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93214)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach, @knz, and @yuzefovich)

___
*[`-- commits` line 39 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/93214#-NIniO4y6r7JHrdhTWFV):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

AFAIK the flow infrastructure is completely unaware of the concept of quiescing, so we - generally speaking - get lucky that the quiescing process takes enough time for all remote flows to perform the cleanup, including performing the deletion of the temp storage directories. Also we probably don't have enough tests that force the spilling to disk (if we don't spill, then there is nothing to delete) so that the cases when we don't get lucky are quite rare. Yet, under stress they can be reproduced pretty easily. It also seems possible that the likelihood of the flakes has recently increased with us fully parallelizing the setup of all flows (until #89649, remote flows had a head start).
</blockquote></details>

Isn't that the answer then - to run flows in tasks? User-level activity like should run in tasks; all RPCs do, for example.
I see we create tasks when starting the flow, but we finish the task to early, without waiting for the slow. I think we should put [this wait](https://github.com/cockroachdb/cockroach/blob/1bbffcc9e82e2cc1731b16bbf838c1316bf6fbc7/pkg/sql/flowinfra/remote_flow_runner.go#L113) into the task.


<!-- Sent from Reviewable.io -->

yuzefovich committed:
flowinfra: create flow runner waiter goroutine as a task

This commit makes the change to create goroutine that is spun up to wait
for the remote flow to finish and to clean up after it as a task on the
stopper. This change allows us to fix the possible race between the server
quiescing (which closes the temp engine) and the vectorized flow cleanup
(which might delete its temp storage directory).

Some care had to be taken to prevent some types of races: on one hand, we
cannot start the flow until the new async waiter task is spun up because
in that case the bug this commit is fixing is still present; on the
other hand, the async task cannot block on `Flow.Wait` method until the
flow is started successfully. Such a setup requires coordination between
the new async task and the flow being started in the current goroutine.

Release note: None

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93214)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @cucaroach, and @knz)

___
*[`-- commits` line 39 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/93214#-NIniO4y6r7JHrdhTWFV):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Isn't that the answer then - to run flows in tasks? User-level activity like should run in tasks; all RPCs do, for example.
I see we create tasks when starting the flow, but we finish the task to early, without waiting for the slow. I think we should put [this wait](https://github.com/cockroachdb/cockroach/blob/1bbffcc9e82e2cc1731b16bbf838c1316bf6fbc7/pkg/sql/flowinfra/remote_flow_runner.go#L113) into the task.
</blockquote></details>

Thanks, that's a good point, done.


<!-- Sent from Reviewable.io -->

cucaroach reviewed (approved):
> Reviewed 2 of 3 files at r3, 2 of 5 files at r5, 5 of 5 files at r9, 1 of 1 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93214)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @knz)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
TFTRs!

bors r+

yuzefovich commented:
Need to adjust a test.

bors r-

yuzefovich reviewed (commented):
I had to make some tweaks to the second commit in order for the error from `Flow.Start` method to be correctly propagated to the client. Would appreciate another look.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93214#-:-NJ7vEKF0U9gBDvgO2OR:bgtqdbk)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @cucaroach, and @knz)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
I'll go ahead and merge since this PR fixes one of the most common flakes right now, but happy to address any feedback in a follow up.

bors r+

yuzefovich commented:
Hm, this doesn't fix the bug anymore 😞 

bors r-
